### ISSUE TITLE
Leaking fds until rlimits reached

### ISSUE THREAD

**Issue by crosbymichael** [prometheus/prometheus#4498]
## Bug Report
**What did you do?**

Run prom with consul discovery

**What did you expect to see?**

Prom not become unresponsive and eat fds.

**What did you see instead? Under which circumstances?**

![image](https://user-images.githubusercontent.com/749551/44085345-35e14696-9f87-11e8-9d15-90e93cacdcdb.png)

**Environment**

* System information:

	`Linux 4.15.0-30-generic x86_64`

* Prometheus version:

	2.3.2

* Prometheus configuration file:

```yaml
global:
  scrape_interval:     30s # Set the scrape interval to every 15 seconds. Default is every 1 minute.
  evaluation_interval: 30s # Evaluate rules every 15 seconds. The default is every 1 minute.
alerting:
  alertmanagers:
  - static_configs:
    - targets:
      # - alertmanager:9093
rule_files:
  # - "first_rules.yml"
  # - "second_rules.yml"

scrape_configs:
        - job_name: 'prometheus'
          static_configs:
          - targets: ['localhost:9090']
        - job_name: 'overwritten-default'
          consul_sd_configs:
          - server:   'yy:8500'
            services: []
            tag: 'metrics'

          relabel_configs:
          - source_labels: ['__meta_consul_service']
            regex:         '(.*)'
            target_label:  'job'
            replacement:   '$1'
          - source_labels: ['__meta_consul_address']
            regex:         '(.*)'
            target_label:  '__meta_consul_service_address'
            replacement:   '$1'
            action: 'replace'
          - source_labels: ['__meta_consul_service_id']
            regex:         '(.*)'
            target_label:  'instance'
            replacement:   '$1'
          - source_labels: ['__meta_consul_service']
            regex: 'containerd'
            target_label: '__metrics_path__'
            replacement: '/v1/metrics'
```

* Logs:
```
Aug 14 05:30:59 aegis-03 boss[2388]: level=error ts=2018-08-14T09:30:59.703921368Z caller=compact.go:432 component=tsdb msg="removed tmp folder after failed compaction" err="open /var/lib/prometheus/01CMVWCBQQS6RTQMR7PVRSD1BR.tmp: too many open files"
Aug 14 05:30:59 aegis-03 boss[2388]: level=error ts=2018-08-14T09:30:59.703982733Z caller=db.go:272 component=tsdb msg="compaction failed" err="persist head block: open chunk writer: open /var/lib/prometheus/01CMVWCBQQS6RTQMR7PVRSD1BR.tmp/chunks: too many open files"
Aug 14 05:31:06 aegis-03 boss[2388]: level=error ts=2018-08-14T09:31:06.309900608Z caller=consul.go:460 component="discovery manager scrape" discovery=consul msg="Error refreshing service" service=metrics tag=metrics err="Get http://j:8500/v1/catalog/service/metrics?stale=&tag=metrics&wait=30000ms: dial tcp: lookup f on 192.168.1.42:53: dial udp 192.168.1.42:53: socket: too many open files"
Aug 14 05:31:36 aegis-03 boss[2388]: level=error ts=2018-08-14T09:31:36.313416194Z caller=consul.go:460 component="discovery manager scrape" discovery=consul msg="Error refreshing service" service=metrics tag=metrics err="Get http://aegis-03.node.aegis:8500/v1/catalog/service/metrics?stale=&tag=metrics&wait=30000ms: dial tcp: lookup v on 192.168.1.42:53: dial udp 192.168.1.42:53: socket: too many open files"
```

**Comment by sipian** [comment]
@crosbymichael
Let me try to replicate this locally. Then I'll look at what's going wrong here.

**Comment by crosbymichael** [comment]
@sipian thanks!

**Comment by sipian** [comment]
@crosbymichael 
I didn't get this error when I set up a consul cluster (a small one though :)).
Can you give me an idea about the number of targets/services that the prometheus-server was scraping?
Also it would we helpful if you could give some details of the consul setup

**Comment by sipian** [comment]
Yeah looks like repeated SIGHUP is causing this.
@gouthamve @brian-brazil Can you confirm that this is the expected behaviour?

**Comment by crosbymichael** [comment]
Well I can tell you that it isn't expected ;) 

Probably just a small bug somewhere on reload not causing old conns or clients to be closed before creating new ones.

**Comment by sipian** [comment]
@brian-brazil 
After doing some digging in [consul.go](https://github.com/prometheus/prometheus/blob/master/discovery/consul/consul.go), I found that new [http.Transport](https://github.com/prometheus/prometheus/blob/master/discovery/consul/consul.go#L171) & [http.Client](https://github.com/prometheus/prometheus/blob/master/discovery/consul/consul.go#L179) structs are made when [NewDiscovery](https://github.com/prometheus/prometheus/blob/master/discovery/consul/consul.go#L162) is called (on reloading config).

And after reloading config, `lsof -p $(pgrep prometheus)` tells that 2 new TCP connections have been established and none of the previous ones to `localhost:8500` have been closed.

I think this might be the culprit.

According to [net/http-godoc](https://golang.org/pkg/net/http/),

> Clients and Transports are safe for concurrent use by multiple goroutines and for efficiency should only be created once and re-used.

[https://github.com/golang/go/issues/4049](https://github.com/golang/go/issues/4049) might be of interest.

**Comment by gouthamve** [comment]
Does this fix it: https://github.com/prometheus/prometheus/commit/1cd29f782cd0643a85c6167740a1ae3bc32abc8d
?

**Comment by sipian** [comment]
Yeah the above commit fixes it in my consul setup.
@crosbymichael Can you test this on your setup?

**Comment by sipian** [comment]
@gouthamve 
I noticed that a similar use of `http.Client` & `http.Transport` is used for triton & marathon service discovery.
Although, I haven't tested them, I suspect these might also have the same problem.

What are your views on this?
If such a behaviour is expected on them too, I will open a PR to fix it.

**Comment by gouthamve** [comment]
Try to replicate it before patching it? In general it is a bad idea to fix code that you haven't run. You could test it, but sadly I don't think any SD mechanisms have tests :(

**Comment by simonpasquier** [comment]
Yes it seems identical to #4425. Hopefully @crosbymichael will confirm that it fixes the problem.

**Comment by sipian** [comment]
@gouthamve 
That's why we have https://github.com/prometheus/prombench/issues/133 :)

**Comment by crosbymichael** [comment]
Thanks, i'll build and deploy master and let you know if that commit fixes it.

**Comment by gouthamve** [comment]
I'm closing this as it is most probably fixed in #4443 Please let me know if it's otherwise.
